ocaml/networkd/lib/network_device_order.ml

```diff
@@ -0,0 +1,526 @@
+(** Generete an order for host network devices and keep the order as stable as possible.
+  *)
+
+module D = Debug.Make (struct let name = __MODULE__ end)
+
+open D
+open Network_interface
+
+let initial_rules_file_path = "/var/lib/xcp/initial_network_device_rules.conf"
+
+let cmd_biosdevname = "/usr/sbin/biosdevname"
+
+(** PCI address in format SBDF: domain:bus:device:function *)
+module Pciaddr = struct
+  type t = Xcp_pci.address
+
+  let default = Xcp_pci.{domain= 0; bus= 0; dev= 0; fn= 0}
+
+  let to_string = Xcp_pci.string_of_address
+
+  exception Parse_error of string
+
+  let of_string_exn s =
+    try Xcp_pci.address_of_string s with _ -> raise (Parse_error s)
+
+  let compare t1 t2 =
+    let open Xcp_pci in
+    match
+      ( compare t1.domain t2.domain
+      , compare t1.bus t2.bus
+      , compare t1.dev t2.dev
+      , compare t1.fn t2.fn
+      )
+    with
+    | x, _, _, _ when x <> 0 ->
+        x
+    | _, x, _, _ when x <> 0 ->
+        x
+    | _, _, x, _ when x <> 0 ->
+        x
+    | _, _, _, x ->
+        x
+end
+
+module PciaddrMap = Map.Make (Pciaddr)
+module MacaddrSet = Set.Make (Macaddr)
+module MacaddrMap = Map.Make (Macaddr)
+module IntMap = Map.Make (Int)
+module IntSet = Set.Make (Int)
+
+module ToMap = struct
```

Comment:
Why is this complicated module needed? I can't easily tell what to_11_map does, even with the comment explaining it.

This seems like jane-street-like levels of overengineering